### PR TITLE
core: add Condition[Memory/CPU]Pressure

### LINKED ISSUE #20139: RFE: skip job activation if not enough system resources (CPU/memory) are available
**Is your feature request related to a problem? Please describe.**
On slow/overloaded/constrained systems it is common to partition system resources (CPU/memory) so that critical tasks are guaranteed to be able to run. This is especially true when memory overcommitting (swap) is disabled. Periodic tasks (eg: timer units) need to have resources carved out too, despite only needing them seldomly.

**Describe the solution you'd like**
Linux makes pressure stall information (PSI) available (https://www.kernel.org/doc/html/latest/accounting/psi.html), which indicate the state of system resources availability within a given window of time, as an indicator of system load.
New unit directives such as `ConditionMemoryPressure=`, `ConditionCPUPressure=` could be made available, which instructs systemd to skip the activation of a unit if the required threshold is not met, for example: `ConditionMemoryPressure=70` following oomd's syntax https://www.freedesktop.org/software/systemd/man/oomd.conf.html#DefaultMemoryPressureLimit=

**Describe alternatives you've considered**
Carving out resources specifically for period tasks, which leads to under-utilization, or carefully building dependency/ordering chains between all timed units, which is unwieldy and not really possible unless a single entity provides them all.

### PR DESCRIPTION
**Pull request by bluca**
By default checks PSI on /proc/pressure, and causes a unit to be skipped
if the threshold is above the given configuration for the avg300
measurement.
Also allow to pass a custom comparison and average, and a particular
slice unit to check under.

Fixes #20139

### CODE REVIEW

File: src/test/test-condition.c [hunk 1]
```diff
@@ -1031,9 +1032,77 @@ static void test_condition_test_os_release(void) {
...
+        condition = condition_new(CONDITION_CPU_PRESSURE, "0.01%", false, false);
+        assert_se(condition);
+        assert_se(condition_test(condition, environ) == 0);
+        condition_free(condition);
+}
+
 int main(int argc, char *argv[]) {
         test_setup_logging(LOG_DEBUG);
 
+if (false) {
```
Comment (keszybz) [inline src/test/test-condition.c]: A leftover ;)

**Review by keszybz — COMMENTED** [review]
The implementation looks good.

But the way that this check is defined seems very _ad-hoc_. In particular, one field uses `some`, the other `full`. And it seems likely that people will request different variants of the check. Thus I wonder if this should not be defined more generically, for example by allowing to specify whether to look at the root cgroup or some slice, and which of the pressure fields to use.

**Comment by bluca** [comment]
> The implementation looks good.
> 
> But the way that this check is defined seems very _ad-hoc_. In particular, one field uses `some`, the other `full`. And it seems likely that people will request different variants of the check. Thus I wonder if this should not be defined more generically, for example by allowing to specify whether to look at the root cgroup or some slice, and which of the pressure fields to use.

Something like: `foo.slice:avg10<30`?
I thought about that, the counterargument is that with this flexibility comes a much more confusing syntax - for scheduling purposes, it seems to me average over 5 minutes would be the most useful metric, the other two are really small windows. Maybe I can try to add support for both formats, if it's just a percentage then assume it's avg300

File: src/shared/condition.c [hunk 2]
```diff
@@ -958,6 +960,38 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+
+        r = parse_permyriad(c->parameter);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse permyriad: %s", c->parameter);
+
+        r = store_loadavg_fixed_point((unsigned long) r / 100, (unsigned long) r % 100, &limit);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse loadavg: %s", c->parameter);
+
+        return pressure.avg300 < limit;
```
Comment (poettering) [inline src/shared/condition.c]: shouldn't this all be "below or equal" rather than "below"?

File: src/test/test-condition.c (hunk 1 above)
Comment (bluca) [reply src/test/test-condition.c]: Whops! And I did double check!

File: src/shared/condition.c (hunk 2 above)
Comment (bluca) [reply src/shared/condition.c]: I guess it depends on how we document it? The manpage says "below" atm

**Comment by bluca** [comment]
Ok, updated: it still allows the simple form "Condition=10%", but also a more complex and complete "Condition=avg10 < 10%". Also can be optionally prefixed with a slice unit name. PTAL!

File: man/systemd.unit.xml [hunk 3]
```diff
@@ -1658,6 +1658,28 @@
...
         </varlistentry>
 
+        <varlistentry>
+          <term><varname>ConditionMemoryPressure=</varname></term>
+          <term><varname>ConditionCPUPressure=</varname></term>
+
+          <listitem><para>Verify that the overall system (memory or CPU) pressure is below the given threshold.
+          This setting takes a colon-separated tuple as argument. The first value is the slice unit under which
+          the pressure will be checked, and it is optional: if not specified, the overall system pressure will be
+          measured, instead of a particular cgroup's.</para>
```
Comment (anitazha) [inline man/systemd.unit.xml]: When there is a condition error (e.g. the provided slice doesn't exist) does the condition succeed or fail? IMO I think the condition should succeed if a slice doesn't exist (since no slice == no pressure)

Also I feel like this implies a dependency on certain systemd slices without providing any kind of actual dependency through the usual unit dependency mechanism. Not sure if that's actually an issue.

File: src/shared/condition.c [hunk 4]
```diff
@@ -958,6 +960,113 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        int r;
+
+        assert(c);
+        assert(c->parameter);
+        assert(IN_SET(c->type, CONDITION_MEMORY_PRESSURE, CONDITION_CPU_PRESSURE));
+
+        if (!is_pressure_supported())
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Pressure Stall Information (PSI) is not supported");
+
+        type = c->type == CONDITION_CPU_PRESSURE ? PRESSURE_TYPE_SOME : PRESSURE_TYPE_FULL;
```
Comment (anitazha) [inline src/shared/condition.c]: cpu.pressure full was added in the past year https://lore.kernel.org/lkml/20201225042248.51737-1-zhouchengming@bytedance.com/. Though it looks like the docs aren't updated. Maybe for consistency switch all of these to "full" or explicitly document that cpu.pressure uses "some"?

Comment (bluca) [reply src/shared/condition.c]: Didn't know that, thanks, will switch to always use 'full' and fallback to 'some' if not available, and document it

File: man/systemd.unit.xml (hunk 3 above)
Comment (bluca) [reply man/systemd.unit.xml]: Due to the way conditions are set up, it would require quite a bit of lifting and refactoring to bring that information down to where it's checked. Also, I figured this allows more interesting patterns, like avoiding trampling over _other_ slices than your own - for example a task in the new desktop background.slice could be set up to be skipped if app.slice is under pressure.
But yes if it's missing it should not be an error, but it should pass.

Comment (bluca) [reply man/systemd.unit.xml]: Changed as requested

File: src/shared/condition.c (hunk 4 above)
Comment (bluca) [reply src/shared/condition.c]: Ok, done, first check 'full' and fallback to 'some' on ENODATA

File: man/systemd.unit.xml [hunk 5]
```diff
@@ -1658,6 +1658,30 @@
...
+
+          <para>The second value is the threshold. It can be specified as a simple percentage value, suffixed with
+          <literal>%</literal>, in which case the pressure will be measured as an average over the last five
+          minutes before the attempt to start the unit is performed, and it will be checked if it is strictly
+          below the specified threshold.</para>
```
Comment (poettering) [inline man/systemd.unit.xml]: Hmm, do we really want to allow the slice specification? and why just slice? why not any unit?

If we allow referring other units we typically would add an implicit dep here, too. But i am not convinced that'd be right for a condition (as it's about silently skipping stuff, i.e. reducing stuff that is run, and pulling in new deps because of that sounds wrong).

I wonder: what is the precise usecase for allowing to specify arbitrary slices here? i.e. why would i want to start a service if some unrelated slice is under pressure or not?

I'd just remove this part I think: either always use the system as a whole or always look at the slice the unit itself is in. Or maybe add a "modifier" to choose between exactly these two slices. But allow arbitrary slices? I don#t see the usecase.

File: man/systemd.unit.xml [hunk 6]
```diff
@@ -1658,6 +1658,30 @@
...
+
+          <para>Alternatively, the average type and the comparison operator can also be specified, for example:
```
Comment (poettering) [inline man/systemd.unit.xml]: alternative to what precisely? the first *and* second arg? or just the latter? please be specific here.

File: src/shared/condition.c [hunk 7]
```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        const char *p, *value;
+        OrderOperator order;
+        int r;
+
+        assert(c);
+        assert(c->parameter);
+        assert(IN_SET(c->type, CONDITION_MEMORY_PRESSURE, CONDITION_CPU_PRESSURE));
+
+        if (!is_pressure_supported())
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Pressure Stall Information (PSI) is not supported");
```
Comment (poettering) [inline src/shared/condition.c]: mmh, to maximize portability of unit files I'd say that on a kernel that lacks PSI we assume the system is always idle.

File: src/shared/condition.c [hunk 8]
```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                value = first;
+        } else {
+                CGroupMask mask, required_mask = c->type == CONDITION_CPU_PRESSURE ? CGROUP_MASK_CPU : CGROUP_MASK_MEMORY;
+                const char *controller = c->type == CONDITION_CPU_PRESSURE ? "cpu.pressure" : "memory.pressure";
+                _cleanup_free_ char *slice_path = NULL;
+
+                r = cg_all_unified();
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to determine whether the unified cgroups hierarchy is used: %m");
+                if (r == 0)
```
Comment (poettering) [inline src/shared/condition.c]: here too: i think we should handle this gracefully and assume the system is not overloaded

File: src/shared/condition.c [hunk 9]
```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Requires the unified cgroups hierarchy");
+
+                r = cg_mask_supported(&mask);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get supported cgroup controllers: %m");
+
+                if (!FLAGS_SET(mask, required_mask))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
```
Comment (poettering) [inline src/shared/condition.c]: ditto

**Comment by poettering** [comment]
So yeah, I am not convinced about allowing arbitrary slices to be specified here. For simplicity I'd dumb it down and only look at the system as a whole

File: src/shared/condition.c [hunk 10]
```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                                               "Cgroup %s controller not available.",
+                                               required_mask == CGROUP_MASK_CPU ? "CPU" : "memory");
+
+                r = cg_slice_to_path(first, &slice_path);
+                if (r < 0)
+                        return log_debug_errno(r, "Cannot determine slice \"%s\" cgroup path: %m", first);
+
+                r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, slice_path, controller, &pressure_path);
```
Comment (poettering) [inline src/shared/condition.c:1029]: hmm, we should probably get PID1's own cgroup, then go one level up from that (i.e. drop the /init.scope/ part in the path). that way we'll look at the top of the current OS' cgroup tree.

File: man/systemd.unit.xml [hunk 11]
```diff
@@ -1658,6 +1658,30 @@
...
+          <literal>avg10 &gt;= 10%</literal>. The supported averages follow what the kernel provides, and are
+          <literal>avg10</literal>, <literal>avg60</literal> and <literal>avg300</literal>. The
```
Comment (poettering) [inline man/systemd.unit.xml]: i am not a fan of exposing avg300 here too much, it's too low-level for my taste. Maybe use our usual way to write times instead, and then a suffix notation?

i.e. 

```
ConditionMemoryPressure=90%/5min
```

File: src/shared/condition.c [hunk 12]
```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        }
+
+        /* If a value including the average type and comparison are passed, parse them,
+         * otherwise we assume just a plain percentage that will be checked if it is
+         * strictly smaller than the current pressure average over 5 minutes. */
+        if (startswith(value, "avg")) {
+                _cleanup_free_ char *key = NULL;
+                const char *e = value;
+
+                r = extract_first_word(&e, &key, "!<=>", EXTRACT_RETAIN_SEPARATORS);
```
Comment (poettering) [inline src/shared/condition.c]: wht's the usecase for allowing all the operators here? i mean it's very unlikely anyone wants to statr more services only when under pressure? so it sounds we can just drop this?

I'd really dumb this down I must say...

File: man/systemd.unit.xml (hunk 5 above)
Comment (bluca) [reply man/systemd.unit.xml]: Pulling a dependency definitely sounds wrong, so I avoided it intentionally.

This was asked here: https://github.com/systemd/systemd/pull/21437#pullrequestreview-810953380
I don't have a use case for this specific knob at the moment, but I think I can see a compelling one: consider the new app.slice/session.slice/background.slice we added for the user session. You could have a timer unit in the background.slice, that also checks the pressure in the app.slice and skips if it's above a threshold. That way you don't just reserve resources, but also ensure that the user is really doing nothing before starting to spin a core (and thus a fan) and so on.

File: src/shared/condition.c (hunk 12 above)
Comment (bluca) [reply src/shared/condition.c]: Well I considered it, but given it comes practically for free, and it's even probably simpler to support it rather than enforce the opposite, so why not? I'm sure someone will find a use for it at some point :-)

File: man/systemd.unit.xml (hunk 11 above)
Comment (bluca) [reply man/systemd.unit.xml]: Given I point to the kernel.org docs I kept the same notation for consistency, but I can change it

File: src/shared/condition.c (hunk 12 above)
Comment (poettering) [reply src/shared/condition.c]: There's big benefit in keeping things simple. I'd not add infra here that has no usecase, and needs extra docs.

Let's be a bit frugal with features: only add what really is useful — we can always extend things later if needed.

File: man/systemd.unit.xml (hunk 11 above)
Comment (poettering) [reply man/systemd.unit.xml]: please do. I really think we should normalize our interfaces to be internally consistent with our own, and put less focus on directly mapping the underlying kernel interfaces.

File: src/shared/condition.c (hunk 10 above)
Comment (bluca) [reply src/shared/condition.c:1029]: I don't follow - the intention here is to check an arbitrary slice

**Comment by bluca** [comment]
> So yeah, I am not convinced about allowing arbitrary slices to be specified here. For simplicity I'd dumb it down and only look at the system as a whole

I think it would be quite useful as an optional parameter, so I've left it in. Applied the rest - now the syntax is `[slice:]percentage[/5min|/1min|/10sec]`

File: man/systemd.unit.xml (hunk 5 above)
Comment (poettering) [reply man/systemd.unit.xml]: but why wouldn't be the root slice for that usecase OK too?

File: src/shared/condition.c (hunk 10 above)
Comment (poettering) [reply src/shared/condition.c:1029]: the root of the cgroupfs and the root of what PID 1 manages are not necessarily the same. They often are, but we actually explicitly support if they are not. This has various usecases. Originally it was mostly about containers, i.e. when cgroup namespacing didn't exist: if we'd start a service manager in some cgroup down the tree it would take possesion of that and nothing else. It's still actively used for the --user stuff, there too we run things in a cgroup down the tree. But we officially allow people to also have some process as PID 1 before system starts that moves systemd down the cgroup tree somewhere, and leaves some sibling for other purposes outside of systemd's control.

File: man/systemd.unit.xml (hunk 5 above)
Comment (bluca) [reply man/systemd.unit.xml]: That would be the sum of the children, no? Ie, if the root is under pressure, it might be because background.slice is, but in this case that's fine, it's the other you care about

File: src/shared/condition.c (hunk 10 above)
Comment (bluca) [reply src/shared/condition.c:1029]: I get that, but I don't understand how to apply it here? It's already looking at the top of the OS tree - `cg_get_path` actually ignores the controller, and simply appends what you pass to `"/sys/fs/cgroup/`. So for example:

`user-1000.slice:100%` returns `/sys/fs/cgroup/user.slice/user-1000.slice/cpu.pressure`
`-.slice:100%` returns `/sys/fs/cgroup/cpu.pressure`
`bork.slice:100%` returns `/sys/fs/cgroup/bork.slice/cpu.pressure`

What's missing?

File: man/systemd.unit.xml (hunk 5 above)
Comment (keszybz) [reply man/systemd.unit.xml]: Yeah, it definitely sounds useful to be able to select an arbitrary slice (or even other unit) here. E.g. on server, start a task if the main payload is not under pressure.

File: src/shared/condition.c [hunk 13]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                        c->type == CONDITION_CPU_PRESSURE ? "cpu" :
+                        "io";
+
+        p = c->parameter;
+        r = extract_many_words(&p, ":", 0, &first, &second, NULL);
+        if (r <= 0)
+                return log_debug_errno(r ?: SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
+        /* If only one parameter is passed, then we look at the global system pressure rather than a specific cgroup. */
+        if (r == 1) {
+                pressure_path = path_join("/proc/pressure", pressure_type);
```
Comment (keszybz) [inline src/shared/condition.c:989]: I'd just say `pressure_path = strjoina("/proc/pressure/", pressure_type);` and save on the oom check.

File: src/shared/condition.c [hunk 14]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+
+        value = strstrip(third);
+        if (!value)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
+
+        r = parse_permyriad(value);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse permyriad: %s", c->parameter);
+
+        r = store_loadavg_fixed_point((unsigned long) r / 100, (unsigned long) r % 100, &limit);
```
Comment (keszybz) [inline src/shared/condition.c]: The points of those casts is to force the type, right? It feels  bit strange to cast the lhs to an unsigned type, and then provide a signed rhs. What about `store_loadavg_fixed_point(r / 100LU, r % 100LU, &limit);` ?

**Review by keszybz — APPROVED** [review]
LGTM. Two minor suggestions, feel free to ignore.

File: src/shared/condition.c (hunk 13 above)
Comment (bluca) [reply src/shared/condition.c:989]: it's so that I can reuse the same variable in the other branch of this `if` block, which is set by `cg_get_path()` so it requires cleanup

File: src/shared/condition.c (hunk 14 above)
Comment (bluca) [reply src/shared/condition.c]: obviously cargo-culted from somewhere else :-) applied as suggested

File: src/shared/condition.c (hunk 13 above)
Comment (keszybz) [reply src/shared/condition.c:989]: Ah, OK, I missed that.

File: man/systemd.unit.xml [hunk 15]
```diff
@@ -1658,6 +1658,29 @@
...
 
+        <varlistentry>
+          <term><varname>ConditionMemoryPressure=</varname></term>
+          <term><varname>ConditionCPUPressure=</varname></term>
+          <term><varname>ConditionIOPressure=</varname></term>
+
+          <listitem><para>Verify that the overall system (memory, CPU or IO) pressure is below or equal to a threshold.
+          This setting takes a colon-separated tuple as argument. The first value is the slice unit under which
+          the pressure will be checked, and it is optional: if not specified, the overall system pressure will be
+          measured, instead of a particular cgroup's.</para>
```
Comment (dnicolodi) [inline man/systemd.unit.xml]: Written like this, I would expect `:10%` to be valid syntax, however it seems that the correct syntax is just `10%`. In other words, the text as it stands says that the setting is always a tuple of value separated by `:` where the first value is optional, thus it is natural to infer that the separator is not optional.

What about something like:  "This setting takes a the threshold value as argument. It can be specified [insert text from second paragraph]. Optionally, the threshold value can be prefixed with the a slice unit under which the pressure will be checked, followed by a colon. If the slice unit is not specified, the overall system pressure will be measured, instead of a particular cgroup's."

Comment (bluca) [reply man/systemd.unit.xml]: updated as suggested

File: man/systemd.unit.xml [hunk 16]
```diff
@@ -1658,6 +1658,29 @@
...
+          five minutes before the attempt to start the unit is performed.
+          Alternatively, the average timespan can also be specified using <literal>/</literal> as a separator, for
+          example: <literal>10%/1min</literal>. The supported timespans match what the kernel provides, and are
+          limited to <literal>10sec</literal>, <literal>1min</literal> and <literal>5min</literal>. The
+          <literal>full</literal> PSI will be checked first, and if not found <literal>some</literal> will be
+          checked. For more details, see the documentation on <ulink
+          url="https://www.kernel.org/doc/html/latest/accounting/psi.html">PSI (Pressure Stall Information)
+          </ulink>.</para>
+
+          <para>Optionally, the threshold value can be prefixed with the a slice unit under which the pressure will be checked,
```
Comment (yuwata) [inline man/systemd.unit.xml]: `the a slice unit`

File: src/shared/condition.c [hunk 17]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        }
+
+        pressure_type = c->type == CONDITION_MEMORY_PRESSURE ? "memory" :
+                        c->type == CONDITION_CPU_PRESSURE ? "cpu" :
+                        "io";
+
+        p = c->parameter;
+        r = extract_many_words(&p, ":", 0, &first, &second, NULL);
+        if (r <= 0)
+                return log_debug_errno(r ?: SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
```
Comment (yuwata) [inline src/shared/condition.c]: r is not a boolean here.

File: src/shared/condition.c [hunk 18]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        loadavg_t *current, limit;
+        ResourcePressure pressure;
+        int r;
+
+        assert(c);
+        assert(c->parameter);
+        assert(IN_SET(c->type, CONDITION_MEMORY_PRESSURE, CONDITION_CPU_PRESSURE, CONDITION_IO_PRESSURE));
+
+        if (!is_pressure_supported()) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Pressure Stall Information (PSI) is not supported, skipping.");
```
Comment (yuwata) [inline src/shared/condition.c]: We usually do not include error cause which will not propagated.

File: src/shared/condition.c [hunk 19]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+
+                slice = strstrip(first);
+                if (!slice)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
+
+                r = cg_all_unified();
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to determine whether the unified cgroups hierarchy is used: %m");
+                if (r == 0) {
+                        log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "PSI condition check requires the unified cgroups hierarchy, skipping.");
```
Comment (yuwata) [inline src/shared/condition.c]: Also here.

File: src/shared/condition.c [hunk 20]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                        return 1;
+                }
+
+                r = cg_mask_supported(&mask);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get supported cgroup controllers: %m");
+
+                if (!FLAGS_SET(mask, required_mask)) {
+                        log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
```
Comment (yuwata) [inline src/shared/condition.c]: And here.

File: src/shared/condition.c [hunk 21]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+
+        p = c->parameter;
+        r = extract_many_words(&p, ":", 0, &first, &second, NULL);
+        if (r <= 0)
+                return log_debug_errno(r ?: SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
+        /* If only one parameter is passed, then we look at the global system pressure rather than a specific cgroup. */
+        if (r == 1) {
+                pressure_path = path_join("/proc/pressure", pressure_type);
+                if (!pressure_path)
+                        return -ENOMEM;
```
Comment (yuwata) [inline src/shared/condition.c]: `log_oom()`?

File: src/shared/condition.c [hunk 22]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                const char *controller = strjoina(pressure_type, ".pressure");
+                _cleanup_free_ char *slice_path = NULL;
+                CGroupMask mask, required_mask;
+                char *slice;
+
+                required_mask = c->type == CONDITION_MEMORY_PRESSURE ? CGROUP_MASK_MEMORY :
+                                c->type == CONDITION_CPU_PRESSURE ? CGROUP_MASK_CPU :
+                                CGROUP_MASK_IO;
+
+                slice = strstrip(first);
```
Comment (yuwata) [inline src/shared/condition.c:1004]: Hm? Maybe `second`, instead of `first`.

File: src/shared/condition.c [hunk 23]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+
+                r = cg_slice_to_path(slice, &slice_path);
+                if (r < 0)
+                        return log_debug_errno(r, "Cannot determine slice \"%s\" cgroup path: %m", slice);
+
+                r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, slice_path, controller, &pressure_path);
+                if (r < 0)
+                        return log_debug_errno(r, "Error getting cgroup pressure path from %s: %m", slice_path);
+
+                value = second;
```
Comment (yuwata) [inline src/shared/condition.c:1033]: And then this should be first?

File: src/shared/condition.c [hunk 24]
```diff
@@ -958,6 +960,133 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+        }
+
+        /* If a value including a specific timespan (in the intervals allowed by the kernel),
+         * parse it, otherwise we assume just a plain percentage that will be checked if it is
+         * smaller or equal to the current pressure average over 5 minutes. */
+        r = extract_many_words(&value, "/", 0, &third, &fourth, NULL);
+        if (r <= 0)
+                return log_debug_errno(r ?: SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
```
Comment (yuwata) [inline src/shared/condition.c]: r is not a boolean.

File: src/shared/condition.c (hunk 22 above)
Comment (yuwata) [reply src/shared/condition.c:1004]: Sorry, I misunderstood the man.

File: src/shared/condition.c [hunk 25]
```diff
@@ -958,6 +960,131 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
...
+                return log_debug_errno(r, "Failed to parse permyriad: %s", c->parameter);
+
+        r = store_loadavg_fixed_point(r / 100LU, r % 100LU, &limit);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse loadavg: %s", c->parameter);
+
+        r = read_resource_pressure(pressure_path, PRESSURE_TYPE_FULL, &pressure);
+        if (r == -ENODATA) /* cpu.pressure 'full' was added recently, fall back to 'some'. */
+                r = read_resource_pressure(pressure_path, PRESSURE_TYPE_SOME, &pressure);
+        if (r == -ENOENT) {
```
Comment (yuwata) [inline src/shared/condition.c:1076]: One question. I guess it is not guaranteed that the path exists when this function is called, if a slice is specified. Is it true?

Comment (keszybz) [reply src/shared/condition.c:1076]: Yes, see the discussion above. We don't want to introduce a dependency on the slice. Instead, if the slice is not present, we assume that it has pressure == 0, and the condition succeeds.

Comment (yuwata) [reply src/shared/condition.c:1076]: I see, thanks.

File: src/shared/condition.c (hunk 10 above)
Comment (Werkov) [reply src/shared/condition.c:1029]: @bluca I'd say the missing part is `cgroup_root` (such as in [unit_default_cgroup_path](https://github.com/systemd/systemd/blob/312dac28695ebe7c316c9b2dec40bc7182f5e28e/src/core/cgroup.c#L1934), alternatively [without the manager](https://github.com/systemd/systemd/blob/312dac28695ebe7c316c9b2dec40bc7182f5e28e/src/basic/cgroup-util.c#L2002)).

> But we officially allow people to also have some process as PID 1 before system starts that moves systemd down the cgroup tree somewhere, and leaves some sibling for other purposes outside of systemd's control.

@poettering Citation needed (I mean where's that documented?)

Comment (bluca) [reply src/shared/condition.c:1029]: how would the result be different? SYSTEMD_CGROUP_CONTROLLER is actually ignored since this only supports cgroups v2

Comment (Werkov) [reply src/shared/condition.c:1029]: The result would be different for the user instance of systemd (or some non-standard tree layouts as implied by @poettering ). For the PID 1, there'd be indeed no difference.

Comment (bluca) [reply src/shared/condition.c:1029]: `cg_get_root_path` returns an empty path both under the system manager and the user instance

Comment (Werkov) [reply src/shared/condition.c:1029]: Oops, I confused it with `manager_setup_cgroup` where it's not called with PID 1 but 0 (self) instead. The latter is what I meant would be missing in the user instance.
IOW, the condition in user instance should: a) without slice -- look at global pressure, b) with a slice specified -- look at user instance slice. Does that make sense?

Comment (bluca) [reply src/shared/condition.c:1029]: see: https://github.com/systemd/systemd/pull/21594
